README.md
# Silkaj
[![Version](https://img.shields.io/pypi/v/silkaj.svg)](https://pypi.python.org/pypi/silkaj) [![License](https://img.shields.io/pypi/l/silkaj.svg)](https://pypi.python.org/pypi/silkaj) [![Python versions](https://img.shields.io/pypi/pyversions/silkaj.svg)](https://pypi.python.org/pypi/silkaj)

- CLI Duniter client written with Python 3.
- [Website](https://silkaj.duniter.org)

## Install
```bash
pip3 install silkaj --user
```

- [Install with Pip](doc/install_pip.md)
- [Install the Development environment](doc/install_poetry.md)
- [Install with the build](doc/install_build.md)
- [Build an executable with Pyinstaller](doc/build_with_pyinstaller.md)

## Usage
- Get help usage with `-h` or `--help` options, then run:
```bash
silkaj <sub-command>
```

- Will automatically request and post data on `duniter.org 443` main Ğ1 node.

- Specify a custom node with `-p` option:
```bash
silkaj -p <address>:<port> <sub-command>
```

## Features
### Currency information
- Currency information
- Display the current Proof of Work difficulty level to generate the next block
- Check the current network
- Explore the blockchain block by block

### Money management
- Send transaction
- Consult the wallet balance

### Web-of-Trust management
- Check sent and received certifications and consult the membership status of any given identity in the Web of Trust
- Check the present currency information stand
- Send certification

### Authentication
- Three authentication methods: Scrypt, file, and (E)WIF

## Wrappers
- [Install as a drop-down for GNOME Shell with Argos](doc/argos.md)
- [How-to: automate transactions and multi-output](doc/how-to_automate_transactions_and_multi-output.md)
- [Transaction generator written in Shell](https://gitlab.com/jytou/tgen)
- [Ğ1Cotis](https://git.duniter.org/matograine/g1-cotis)
- [G1pourboire](https://git.duniter.org/matograine/g1pourboire)
- [Ğ1SMS](https://git.duniter.org/clients/G1SMS/)
- [Ğmixer](https://git.duniter.org/tuxmain/gmixer-py/)

### Dependencies
Silkaj is based on Python dependencies:

- [Click](https://click.palletsprojects.com/): Command Line Interface Creation Kit.
- [DuniterPy](https://git.duniter.org/clients/python/duniterpy/): Python APIs library to implement duniter clients softwares.
- [Tabulate](https://bitbucket.org/astanin/python-tabulate/overview): to display charts.

### Names
I wanted to call that program:
- bamiyan
- margouillat
- lsociety
- cashmere

I finally called it `Silkaj` as `Silk` in esperanto.

### Website
- [Silkaj website sources](https://git.duniter.org/websites/silkaj_website/)

## Packaging status
[![Packaging status](https://repology.org/badge/vertical-allrepos/silkaj.svg)](https://repology.org/project/silkaj/versions)


CHANGELOG.md
## v0.9.0 (17th April 2021)

### [Milestone v0.9.0](https://git.duniter.org/clients/python/silkaj/milestones/13)

- #345, #385, !171: Refactor `id`/`lookup` command exclusively using `/wot/lookup`
  - Display non-member uids when passing a pubkey
  - Use same algorithm as `choose_identity()` uses
  - Rename `id` to `lookup` command
- #377, !172: `balance`: Fix undefined variable in case of 'Total' label

Plus what can be found bellow in v0.9.0rc


## v0.9.0rc (24th March 2021)
### Code
#### `tx`
- #281, !129: Handle transaction size limit properly
- #257, #312, #356: Handle chained transactions/Change txs lost while sending big amount
- #296, #362, !154, !157: Prevent sending transaction with 0 as amounts
- #172, !165: Refactor tx confirmation, by using `click.confirm()`

#### `balance`
- #300, !164: `balance`: Display corresponding member identity uid
- #366, !159: Fix wrong `DuniterError` exception handling in `wot.identity_of`
- #377, !166: `balance`: Document `money.show_amount_from_pubkey()`
- #342, !151: Don’t allow to pass multiple times the same pubkey to the `balance` command

#### Others
- #218, !160: `history`: Add option to display the complete pubkeys
- #314, !165: Display option for `cert`, `membership` commands
- !165: Make `--dry-run` option a generic one
- #378, !165: Create and use generic `send_doc_confirmation()` in `cert` and `membership` commands
- #176, !149: Get rid of `PyNaCl` and use `base58` module
- #309, !163: `wot`: Fix legend about received certifications
- #208: `argos`: Remove duplicate call to `CurrencySymbol`

#### Tests
- #213, !130: Write unit tests for the `tx` command
- #282, !130: Split `patched.py` into files
- #335, !130: Merge the two functions testing `transaction_amount()`
- #363, !129: Returns balance from `patched_get_source()`
- #368, !161: Assertions are not tested when testing system exit
- #362, !156: Change "moul" id in tests

### Meta
- #240, !150: Drop Python 3.5 support
- #294, !150, docker/python3/poetry!1: Add support and set-up Python 3.9 test job
- #270: Silkaj v0.8.1 package for Debian Bullseye (v11)
- #226, !158: Get rid of `ipaddress` dependency
- #290, !162: Update `pre-commit` dev dependency
- #267: Update the copyright date to 2021 in the headers of every source files
- !150, !155, !167, Update DuniterPy from v0.58.1 to v0.62.0
- #313, !148: Be compatible with and handle new features from Poetry v1.1
- #299, !147: Introduce dev version suffix

---
Thanks @matograine, @moul, @atrax

## v0.8.1 (30th November 2020)
- #358, !152: Update DuniterPy to v0.58.1, to support `libnacl` v1.7.2

Thanks @matograine

## v0.8.0 (18th October 2020)

### [Milestone v0.8.0](https://git.duniter.org/clients/python/silkaj/milestones/8)

### Code
#### Transaction
- #111, !108: Support passing different amounts on multi-recipients tx
  - **Breaking change**: Rename `--output` option to `--recipient`
- Add extra small options to ease passing multiple amounts and recipients:
  - `-a/--amount`
  - `-d/--amountUD`
  - `-r/--recipient`
  - `-c/--comment`
- Add possibility to pass multiple options:
  - **Breaking change**: recipients public keys are no longer `:` separated: `-r A -r B`

- #232, !131, !132: Identities not retrieved for tx with several issuers, and to display the tx history
- #236, !107: Improve the confirmation display
- !144: Rework confirmation fields titles
- #235: Make sure only one option is passed to retrieve the amount of the transaction

#### Membership, WoT
- #88, !140: Add `membership` command
- #88, !144: Rework table fields names
- #140, !140: Ability to pass an `uid` or a `pubkey` to `wot`, and `cert`, `membership` commands
  - Implement identity choice selector

#### Checksum
- #237, !132: **Breaking change**: Switch back the checksum delimiter from `!` to `:`
- #323, !132: Handle pubkey's checksum in the tx code
- #301, !143: Generalize pubkey checksum display and verifiction, Add `chekcsum` command
- #320, !143: Incorrect use of `check_public_key()` in `id` command

#### Others
- #262, !123: Add new `verify` command to check blocks’ signatures
- #264, !133: Disable the broken `net` command
- !131: Display `powMin` in a row in the `blocks` explorer
- #210, !115: Close client session in every cases
- #223: Make Click context optional to be able to call functions from an external module
- #255, !113: `balance`: display the content in tables
- #269, !133: Move `convert_time()` to `tui.py`
- #278, !128: Fix PubSec regex
- #336, !141: `history`: Pubkeys display issue with multisig txs

### Dev Env
#### Poetry migration
- #182: Migrate from Pipenv and `setup.py` to Poetry
- #249: Install Poetry stable when v1 is released
- #263, !127: Post migration tasks (black, poetry)
- #276, !120: Pip installation do not install `silkaj` executalbe into `$HOME/.local/bin`

#### CI/CD set-up
- #245: Automated containers builds with Poetry installed for Python versions 3.5, 3.6, 3.7, and 3.8
- #149: CI/CD set up
- #105: Deploy on PyPI from GitLab CD
- #146: Add a coverage badge
- #284, !124: `build` and `tests` jobs are not retriggered in case of source code change
- #286, !126: Use latest Black version from PyPI in the container
- !131: Use `rules` instead of `only/except`

#### Tests
- #241: Can not run test with Click utility

### Dependencies
- #259: `attr` error while installing with `pip`
- !121, !131, !142: Update DuniterPy from v0.55.1 to v0.58.0
- #251, !140: Introduce `pendulum` date utility
- Introduce `pytest-sugar`
- Update PyNaCl to v1.4.0
- Update Click to v7.1.2
- #338, !140: Update black to v20

### Python versions support
We added the support for Python 3.8.
#240: It is planned that v0.8.x versions are going to be the last releases with Python 3.5 support
since [its support from the Python project has been dropped September 30th of 2020](https://pythoninsider.blogspot.com/2020/10/python-35-is-no-longer-supported.html).

### Documentation
- #202: Document contribution process in `CONTRIBUTING.md`

silkaj/net.py
"""
Copyright  2016-2021 Maël Azimi <m.a@moul.re>

Silkaj is free software: you can redistribute it and/or modify
it under the terms of the GNU Affero General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

Silkaj is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
GNU Affero General Public License for more details.

You should have received a copy of the GNU Affero General Public License
along with Silkaj. If not, see <https://www.gnu.org/licenses/>.
"""

"""
from click import command, option, get_terminal_size
from datetime import datetime
from os import system
from collections import OrderedDict
from tabulate import tabulate
from asyncio import sleep

from duniterpy.api.client import Client
from duniterpy.api import bma

from silkaj.tools import coroutine
from silkaj.wot import identity_of
from silkaj.network_tools import (
    discover_peers,
    best_endpoint_address,
    ClientInstance,
)
from silkaj.tools import message_exit
from silkaj.tui import convert_time
from silkaj.constants import ASYNC_SLEEP


def get_network_sort_key(endpoint):
    t = list()
    for akey in network_sort_keys:
        if akey == "diffi" or akey == "block" or akey == "port":
            t.append(int(endpoint[akey]) if akey in endpoint else 0)
        else:
            t.append(str(endpoint[akey]) if akey in endpoint else "")
    return tuple(t)


@command("net", help="Display network view")
@option(
    "--discover", "-d", is_flag=True, help="Discover the network (could take a while)"
)
@option(
    "--sort",
    "-s",
    default="block,member,diffi,uid",
    show_default=True,
    help="Sort column names comma-separated",
)
@coroutine
async def network_info(discover, sort):
    global network_sort_keys
    network_sort_keys = sort.split(",")
    width = get_terminal_size()[0]
    if width < 146:
        message_exit(
            "Wide screen need to be larger than 146. Current width: " + str(width)
        )
    # discover peers
    # and make sure fields are always ordered the same
    infos = [
        OrderedDict(
            (i, p.get(i, None)) for i in ("domain", "port", "ip4", "ip6", "pubkey")
        )
        for p in await discover_peers(discover)
    ]
    client = ClientInstance().client
    diffi = await client(bma.blockchain.difficulties)
    members = 0
    print("Getting informations about nodes:")
    for i, info in enumerate(infos):
        ep = info
        api = "BASIC_MERKLED_API " if ep["port"] != "443" else "BMAS "
        api += ep.get("domain") + " " if ep["domain"] else ""
        api += ep.get("ip4") + " " if ep["ip4"] else ""
        api += ep.get("ip6") + " " if ep["ip6"] else ""
        api += ep.get("port")
        print("{0:.0f}%".format(i / len(infos) * 100, 1), end=" ")
        best_ep = best_endpoint_address(info, False)
        print(best_ep if best_ep is None else info[best_ep], end=" ")
        print(info["port"])
        await sleep(ASYNC_SLEEP)
        try:
            info["uid"] = await identity_of(info["pubkey"])
            info["uid"] = info["uid"]["uid"]
            info["member"] = "yes"
            members += 1
        except:
            info["uid"] = None
            info["member"] = "no"
        info["pubkey"] = info["pubkey"][:5] + "…"
        for d in diffi["levels"]:
            if info.get("uid") is not None:
                if info["uid"] == d["uid"]:
                    info["diffi"] = d["level"]
                if len(info["uid"]) > 10:
                    info["uid"] = info["uid"][:9] + "…"
        sub_client = Client(api)
        current_blk = await sub_client(bma.blockchain.current)
        if current_blk is not None:
            info["gen_time"] = convert_time(current_blk["time"], "hour")
            if width > 171:
                info["mediantime"] = convert_time(current_blk["medianTime"], "hour")
            if width > 185:
                info["difftime"] = convert_time(
                    current_blk["time"] - current_blk["medianTime"], "hour"
                )
            info["block"] = current_blk["number"]
            info["hash"] = current_blk["hash"][:10] + "…"
            summary = await sub_client(bma.node.summary)
            info["version"] = summary["duniter"]["version"]
        await sub_client.close()
        if info.get("domain") is not None and len(info["domain"]) > 20:
            info["domain"] = "…" + info["domain"][-20:]
        if info.get("ip6") is not None:
            if width < 156:
                info.pop("ip6")
            else:
                info["ip6"] = info["ip6"][:8] + "…"
    await client.close()
    print(
        len(infos),
        "peers ups, with",
        members,
        "members and",
        len(infos) - members,
        "non-members at",
        datetime.now().strftime("%H:%M:%S"),
    )
    infos = sorted(infos, key=get_network_sort_key)
    print(tabulate(infos, headers="keys", tablefmt="orgtbl", stralign="center"))

"""
